### PR TITLE
Add GKE NetworkPolicy support

### PR DESCRIPTION
**Pull request by kahun**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Add NetworkPolicy support in GCPManagedControlPlane object

**TODOs**:
- [X] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:
```release-note
Add NetworkPolicy support in GCPManagedControlPlane object
```

### CODE REVIEW

**Review by cpanato — COMMENTED** [review]
/ok-to-test

File: cloud/services/container/clusters/reconcile.go [hunk 1]
```diff
@@ -268,6 +268,9 @@ func (s *Service) createCluster(ctx context.Context, log *logr.Logger) error {
 	if !s.scope.IsAutopilotCluster() {
 		cluster.NodePools = scope.ConvertToSdkNodePools(nodePools, machinePools, isRegional, cluster.Name)
 	}
+	if s.scope.GCPManagedControlPlane.Spec.NetworkPolicy != nil {
+		cluster.NetworkPolicy = infrav1exp.ConvertToSdkNetworkPolicy(s.scope.GCPManagedControlPlane.Spec.NetworkPolicy)
```
Comment (richardcase) [inline cloud/services/container/clusters/reconcile.go:272]: The "convert" functions are currently located in the same package as this code or in the "scope" package. Could we follow existing conventions and move the convert function into either of those 2 places?

File: exp/api/v1beta1/types.go [hunk 2]
```diff
@@ -144,3 +144,21 @@ func ConvertToSdkLinuxNodeConfig(linuxNodeConfig *LinuxNodeConfig) *containerpb.
 	}
 	return &sdkLinuxNodeConfig
 }
+
+// ConvertToSdkNetworkPolicy converts NetworkPolicy config to a value that is used by GCP SDK.
+func ConvertToSdkNetworkPolicy(networkPolicy *NetworkPolicy) *containerpb.NetworkPolicy {
```
Comment (richardcase) [inline exp/api/v1beta1/types.go:149]: Following on from a previous comment can this be moved to a different package.

File: exp/api/v1beta1/gcpmanagedcontrolplane_types.go [hunk 3]
```diff
@@ -142,6 +146,25 @@ type MasterAuthorizedNetworksConfigCidrBlock struct {
 	CidrBlock string `json:"cidr_block,omitempty"`
 }
 
+// NetworkPolicy represents configuration options for NetworkPolicy feature of the GKE cluster.
+type NetworkPolicy struct {
```
Comment (richardcase) [inline exp/api/v1beta1/gcpmanagedcontrolplane_types.go:150]: If a user supplies a **NetworkPolicy** should at least one of these 2 fields be specified?

File: exp/api/v1beta1/gcpmanagedcontrolplane_webhook.go [hunk 4]
```diff
@@ -130,6 +130,13 @@ func (r *GCPManagedControlPlane) ValidateUpdate(oldRaw runtime.Object) (admissio
 		)
 	}
 
+	if !cmp.Equal(r.Spec.NetworkPolicy, old.Spec.NetworkPolicy) {
```
Comment (richardcase) [inline exp/api/v1beta1/gcpmanagedcontrolplane_webhook.go:133]: With network polices can you do this:

- create the cluster without a network policy and then enable a policy at a later date

Just wanted to double check that.

File: cloud/services/container/clusters/reconcile.go (hunk 1 above)
Comment (kahun) [reply cloud/services/container/clusters/reconcile.go]: Done

File: exp/api/v1beta1/types.go (hunk 2 above)
Comment (kahun) [reply exp/api/v1beta1/types.go]: Done

File: exp/api/v1beta1/gcpmanagedcontrolplane_types.go [hunk 5]
```diff
@@ -142,6 +146,25 @@ type MasterAuthorizedNetworksConfigCidrBlock struct {
...
+	// The selected network policy provider.
+	// +optional
+	Provider *NetworkPolicyProvider `json:"provider,omitempty"`
+	// Whether network policy is enabled on the cluster.
+	// +optional
+	Enabled *bool `json:"enabled,omitempty"`
```
Comment (richardcase) [inline exp/api/v1beta1/gcpmanagedcontrolplane_types.go:156]: I don't think we need **Enabled** in the API. If someone wants to enable NetworkPolicy i think having a non-nil **NetworkPolicy** is sufficient.

File: exp/api/v1beta1/gcpmanagedcontrolplane_types.go [hunk 6]
```diff
@@ -142,6 +146,25 @@ type MasterAuthorizedNetworksConfigCidrBlock struct {
 	CidrBlock string `json:"cidr_block,omitempty"`
 }
 
+// NetworkPolicy represents configuration options for NetworkPolicy feature of the GKE cluster.
+type NetworkPolicy struct {
+	// The selected network policy provider.
+	// +optional
+	Provider *NetworkPolicyProvider `json:"provider,omitempty"`
```
Comment (richardcase) [inline exp/api/v1beta1/gcpmanagedcontrolplane_types.go:153]: We could ptobably add the kubebilder enum annotation here.

File: exp/api/v1beta1/gcpmanagedcontrolplane_types.go [hunk 7]
```diff
@@ -142,6 +146,14 @@ type MasterAuthorizedNetworksConfigCidrBlock struct {
 	CidrBlock string `json:"cidr_block,omitempty"`
 }
 
+// NetworkPolicy represents configuration options for NetworkPolicy feature of the GKE cluster.
+type NetworkPolicy struct {
+	// The selected network policy provider.
+	// +kubebuilder:validation:Enum=calico
+	// +optional
+	Provider string `json:"provider,omitempty"`
```
Comment (richardcase) [inline exp/api/v1beta1/gcpmanagedcontrolplane_types.go:154]: If NetworkPolicy is supplied by the user should we require that they supply **Provider**. It feels like they should.

**Review by richardcase — COMMENTED** [review]
This is looking good, just one comment from my side.

**Comment by iamjanr** [comment]
Hi @richardcase.
This PR is on lgtm pendig, could you it be merged?

**Pull request closed without merging**